galley/pkg/config/analysis/analyzers/destinationrule/targets.go

```diff
@@ -0,0 +1,214 @@
+					msg.NewConflictingDestinationRulesHost(r, errorEntries, drHost, hs.subset))
+			}
+			for _, r := range wildCardEntries {
+				// We need to create the duplicate report here of wildcard subsets since searching the other way isn't possible
+				ctx.Report(metadata.IstioNetworkingV1Alpha3Destinationrules,
+					msg.NewConflictingDestinationRulesHost(r, errorEntries, drHost, "~"))
+			}
+		}
+
+		nsScopedFqdn := util.NewScopedFqdn(drNamespace, drNamespace, drHost)
+		allNsScopedFqdn := util.NewScopedFqdn(util.ExportToAllNamespaces, drNamespace, drHost)
+
+		_, nsFound := seHosts[nsScopedFqdn]
+		_, allFound := seHosts[allNsScopedFqdn]
+		targetService, serviceFound := services[nsScopedFqdn]
+
+		if !nsFound && !allFound && !serviceFound {
+			// DestinationRule host does not match any ServiceEntry or platform Service
+			for _, r := range resources {
+				ctx.Report(metadata.IstioNetworkingV1Alpha3Destinationrules,
+					msg.NewReferencedResourceNotFound(r, "host", drHost))
+			}
+		}
+
+		if serviceFound {
+			// Check subset labels
+			for _, r := range resources {
+				dr := r.Item.(*v1alpha3.DestinationRule)
+			NextSubset:
+				for _, ss := range dr.GetSubsets() {
+					if ss.Name != hs.subset {
+						continue
+					}
+					subsetSelector := k8s_labels.SelectorFromSet(k8s_labels.Set(ss.GetLabels()))
+					for _, targetPodLabels := range targetService.podLabels {
+						if subsetSelector.Matches(targetPodLabels) {
+							continue NextSubset
+						}
+					}
+					// No target pods found
+					errDest := fmt.Sprintf("%s/subset:%s", drHost, ss.Name)
+					ctx.Report(metadata.IstioNetworkingV1Alpha3Destinationrules,
+						msg.NewReferencedResourceNotFound(r, "subset", errDest))
+				}
+			}
+		}
+	}
+}
+
+func hostSubsetCombinations(ctx analysis.Context) map[hostAndSubset][]*resource.Entry {
+	seenCombinations := map[hostAndSubset][]*resource.Entry{}
+
+	addMatch := func(hs hostAndSubset, r *resource.Entry) {
+		if subset, found := seenCombinations[hs]; found {
+			subset = append(subset, r)
+			seenCombinations[hs] = subset
+		} else if !found {
+			seenCombinations[hs] = []*resource.Entry{r}
+		}
+	}
+
+	ctx.ForEach(metadata.IstioNetworkingV1Alpha3Destinationrules, func(r *resource.Entry) bool {
+		dr := r.Item.(*v1alpha3.DestinationRule)
+		drNamespace, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+
+		// If there are no subsets, then we need to match ALL the subsets..
+		if len(dr.GetSubsets()) < 1 || (len(dr.GetSubsets()) == 1 && dr.GetSubsets()[0].Name == "~") {
+			hs := hostAndSubset{
+				host:   util.GetResourceNameFromHost(drNamespace, dr.GetHost()),
+				subset: util.AllSubsets,
+			}
+			addMatch(hs, r)
+		}
+		for _, ss := range dr.GetSubsets() {
+			hs := hostAndSubset{
+				host:   util.GetResourceNameFromHost(drNamespace, dr.GetHost()),
+				subset: ss.GetName(),
+			}
+			addMatch(hs, r)
+		}
+
+		return true
+	})
+
+	return seenCombinations
+}
+
+type servicePod struct {
+	serviceSelector k8s_labels.Selector
```

Comment:
Do we need to store serviceSelector? We don't use it outside the availableServices function, right? 

Why not just have availableServices return a map of service -> podLabels?